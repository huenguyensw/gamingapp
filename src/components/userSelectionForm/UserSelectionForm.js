import React, { useState } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faHandPaper, faHandRock } from '@fortawesome/free-solid-svg-icons';
import { faHandScissors } from '@fortawesome/free-regular-svg-icons';




const UserSelectionForm = ({ playerName, choiced, setChoiceOfPlayer, choiceOfPlayer, results, setResults, setUpdatedResult, gameMode, currentResult, setWinner }) => {
  
  
  const handleClick = (val) => {
    
    setChoiceOfPlayer(val);
    currentResult.current.push({name: playerName, choice: val});
    results[playerName] = (results[playerName] || []).concat([val]);
    setResults(results);
    
    choiced.current = choiced.current + 1;
    if (choiced.current == 2 && gameMode == 1) {
      setUpdatedResult(true);
      
      if((currentResult.current[0].choice == currentResult.current[1].choice - 1)|| (currentResult.current[1].choice == currentResult.current[0].choice - 2)){
        setWinner(`${currentResult.current[0].name} won!`)
      } else if((currentResult.current[1].choice == currentResult.current[0].choice - 1)|| (currentResult.current[0].choice == currentResult.current[1].choice - 2)){
        setWinner(`${currentResult.current[1].name} won!`)
      } else {
        setWinner('Draw!')
      }
      currentResult.current = [];
      choiced.current = 0;
    }
    if (gameMode == 2 && choiced.current == 1) {
      choiced.current = choiced.current + 1;
      const randomInt = Math.floor(Math.random() * 3);
      results['computer'] = (results['computer'] || []).concat([randomInt]);
      currentResult.current.push({name: 'computer', choice: randomInt});
      if((currentResult.current[0].choice == currentResult.current[1].choice - 1)|| (currentResult.current[1].choice == currentResult.current[0].choice - 2)){
        setWinner(`${currentResult.current[0].name} won!`)
      } else if((currentResult.current[1].choice == currentResult.current[0].choice - 1)|| (currentResult.current[0].choice == currentResult.current[1].choice - 2)){
        setWinner(`Regretfully! ${currentResult.current[0].name}, you have lost this time. Try again!`)
      }
      else {
        setWinner('Draw!')
      }
      
      setResults(results);
      setUpdatedResult(true);
      choiced.current = 0;
      currentResult.current = [];
    }
  }

  return (
      <div className='player-section'>
          <h2>{playerName}</h2>
          {playerName == 'computer'
            ? <section data-testid='computerContent'>Picked random</section>
            : choiceOfPlayer == null
              ? <section data-testid='playingIcons' className='playing-icons'>
                <FontAwesomeIcon  data-testid='rockIcon' icon={faHandRock} size='3x' className='rock' onClick={() => handleClick(0)} />
                <FontAwesomeIcon data-testid='paperIcon' icon={faHandPaper} size='3x' className='paper' onClick={() => handleClick(2)} />
                <FontAwesomeIcon data-testid='scissorIcon' icon={faHandScissors} size='3x' className='scissor' onClick={() => handleClick(1)} />
              </section>
              : <p>You picked</p>}
      </div>
  )
}

export default UserSelectionForm
