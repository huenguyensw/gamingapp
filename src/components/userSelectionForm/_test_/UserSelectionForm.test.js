import { render, screen, fireEvent } from "@testing-library/react";
import UserSelectionForm from "../UserSelectionForm";


describe("userSelectionForm", () => {
    const setChoiceOfPlayer = jest.fn();
    const setResults = jest.fn();
    const setUpdatedResult = jest.fn();
    const setWinner = jest.fn();

    test('should render playerName', () => {
        //Given
        const playerName = 'Hanna';
        //When
        render(<UserSelectionForm playerName={playerName} />);
        const playerNameElement = screen.getByText(playerName)

        //Then
        expect(playerNameElement).toBeInTheDocument()
    })

    test("text 'Picked random' should be rendered when player is 'computer'", () => {
        //Given
        const playerName = 'computer';

        //When
        render(<UserSelectionForm playerName={playerName} />);

        //solution 2
        const playerNameElement = screen.getByTestId('computerContent')

        //Then
        expect(playerNameElement).toHaveTextContent('Picked random')
    })

    it('should render playing icons when choiceOfPlayer is null', () => {
        //Given
        const playerName = "Player 1";
        const choiceOfPlayer = null;

        //When
        render(<UserSelectionForm playerName={playerName} choiceOfPlayer={choiceOfPlayer} />)
        const iconSectionElement = screen.getByTestId('playingIcons')

        //Then
        expect(iconSectionElement).toBeInTheDocument();
    });

    it("should render 'You picked' after player made a choice",()=>{
        //Given
        const playerName = "Player 1";
        const choiceOfPlayer = 1;

        //When
        render(<UserSelectionForm playerName={playerName} choiceOfPlayer={choiceOfPlayer} />)
        const iconSectionElement = screen.getByText('You picked')

        //Then
        expect(iconSectionElement).toBeInTheDocument();
    })

    it('should call the setChoiceOfPlayer function when player makes a choice',()=>{
        const playerName = 'Player 1';
        const results = {};
        const currentResult = {current:[]};
        const choiced = {current: 0}
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setChoiceOfPlayer).toHaveBeenCalledWith(2);
    })

    it('Human vs Computer: should call the setUpdatedResult function after player makes a choice',()=>{
        const playerName = 'Player 1';
        const results = {};
        const choiced = {current: 0};
        const currentResult = {current:[]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={2} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setUpdatedResult).toHaveBeenCalledWith(true);
    })

    it('humanvs computer: should call the setResults function two times after player make a choice',()=>{
        const playerName = 'Player 1';
        const results = {};
        const choiced = {current: 0};
        const currentResult = {current:[]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={2} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setResults).toHaveBeenCalledTimes(2);
    })

    it('human vs computer: should call the setWinner function after player make a choice',()=>{
        const playerName = 'Player 1';
        const results = {};
        const choiced = {current: 0};
        const currentResult = {current:[]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={2} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setWinner).toHaveBeenCalledTimes(1);
    })

    it('Human vs Human: should call the setUpdatedResult function after both players made their choice',()=>{
        const playerName = 'Player 2';
        const results = {};
        const choiced = {current: 1};
        const currentResult = {current:[{name: playerName,choice: 2}]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={1} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setUpdatedResult).toHaveBeenCalledWith(true);
    })

    it('human vs human: should call the setResults function when player make a choice',()=>{
        const playerName = 'Player 1';
        const results = {};
        const choiced = {current: 0};
        const currentResult = {current:[]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={1} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setResults).toHaveBeenCalledTimes(1);
        expect(setResults).toHaveBeenCalledWith({"Player 1": [2]});
    })

    
    it('human vs human: should call the setWinner function after both players made their choices',()=>{
        const playerName = 'Player 1';
        const results = {};
        const choiced = {current: 1};
        const currentResult = {current:[{name: playerName, choice: 2}]};
        render(<UserSelectionForm playerName = {playerName} choiceOfPlayer={null} results={results} currentResult={currentResult} setChoiceOfPlayer={setChoiceOfPlayer} setResults={setResults} choiced={choiced} setUpdatedResult={setUpdatedResult}
        gameMode={1} setWinner={setWinner}/>)
        const iconElement = screen.getByTestId('paperIcon');
        fireEvent.click(iconElement);
        expect(setWinner).toHaveBeenCalledTimes(1);
    })
})