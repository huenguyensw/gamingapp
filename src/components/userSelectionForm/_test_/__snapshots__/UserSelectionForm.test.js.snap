// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userSelectionForm Should render same snapshot 1`] = `
<DocumentFragment>
  <div
    class="player-section"
  >
    <h2>
      Hanna
    </h2>
    <section
      class="playing-icons"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-hand-back-fist fa-3x rock"
        data-icon="hand-back-fist"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M144 0C117.5 0 96 21.5 96 48V96v28.5V176c0 8.8-7.2 16-16 16s-16-7.2-16-16V149.3l-9 7.5C40.4 169 32 187 32 206V244c0 38 16.9 74 46.1 98.3L128 384v96c0 17.7 14.3 32 32 32H320c17.7 0 32-14.3 32-32V374.7c46.9-19 80-65 80-118.7V176 160 144c0-26.5-21.5-48-48-48c-12.4 0-23.6 4.7-32.1 12.3C350 83.5 329.3 64 304 64c-12.4 0-23.6 4.7-32.1 12.3C270 51.5 249.3 32 224 32c-12.4 0-23.6 4.7-32.1 12.3C190 19.5 169.3 0 144 0z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-hand fa-3x paper"
        data-icon="hand"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M288 32c0-17.7-14.3-32-32-32s-32 14.3-32 32V240c0 8.8-7.2 16-16 16s-16-7.2-16-16V64c0-17.7-14.3-32-32-32s-32 14.3-32 32V336c0 1.5 0 3.1 .1 4.6L67.6 283c-16-15.2-41.3-14.6-56.6 1.4s-14.6 41.3 1.4 56.6L124.8 448c43.1 41.1 100.4 64 160 64H304c97.2 0 176-78.8 176-176V128c0-17.7-14.3-32-32-32s-32 14.3-32 32V240c0 8.8-7.2 16-16 16s-16-7.2-16-16V64c0-17.7-14.3-32-32-32s-32 14.3-32 32V240c0 8.8-7.2 16-16 16s-16-7.2-16-16V32z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-hand-scissors fa-3x scissor"
        data-icon="hand-scissors"
        data-prefix="far"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M.2 276.3c-1.2-35.3 26.4-65 61.7-66.2l3.3-.1L57 208.1C22.5 200.5 .7 166.3 8.3 131.8S50.2 75.5 84.7 83.2l173 38.3c2.3-2.9 4.7-5.7 7.1-8.5l18.4-20.3C299.9 74.5 323.5 64 348.3 64l10.2 0c54.1 0 104.1 28.7 131.3 75.4l1.5 2.6c13.6 23.2 20.7 49.7 20.7 76.6L512 344c0 66.3-53.7 120-120 120l-8 0-96 0c-35.3 0-64-28.7-64-64c0-2.8 .2-5.6 .5-8.3c-19.4-11-32.5-31.8-32.5-55.7c0-.8 0-1.6 0-2.4L66.4 338c-35.3 1.2-65-26.4-66.2-61.7zm63.4-18.2c-8.8 .3-15.7 7.7-15.4 16.5s7.7 15.7 16.5 15.4l161.5-5.6c9.8-.3 18.7 5.3 22.7 14.2s2.2 19.3-4.5 26.4c-2.8 2.9-4.4 6.7-4.4 11c0 8.8 7.2 16 16 16c9.1 0 17.4 5.1 21.5 13.3s3.2 17.9-2.3 25.1c-2 2.7-3.2 6-3.2 9.6c0 8.8 7.2 16 16 16l96 0 8 0c39.8 0 72-32.2 72-72l0-125.4c0-18.4-4.9-36.5-14.2-52.4l-1.5-2.6c-18.6-32-52.8-51.6-89.8-51.6l-10.2 0c-11.3 0-22 4.8-29.6 13.1l-17.5-15.9 17.5 15.9-18.4 20.3c-.6 .6-1.1 1.3-1.7 1.9l57 13.2c8.6 2 14 10.6 12 19.2s-10.6 14-19.2 12l-85.6-19.7L74.3 130c-8.6-1.9-17.2 3.5-19.1 12.2s3.5 17.2 12.2 19.1l187.5 41.6c10.2 2.3 17.8 10.9 18.7 21.4l.1 1c.6 6.6-1.5 13.1-5.8 18.1s-10.6 7.9-17.2 8.2L63.6 258.1z"
          fill="currentColor"
        />
      </svg>
    </section>
  </div>
</DocumentFragment>
`;
